tests/unit/runtime/zero/test_zero_nesting_init.py

```diff
@@ -20,6 +20,16 @@ def test_nesting_init(self):
             with deepspeed.zero.Init(config_dict_or_path=ds_config):
                 model = torch.nn.Linear(4, 4)
 
-        deepspeed_engine, *_ = deepspeed.initialize(model=model, config_params=ds_config)
         # ensure that zero3 processed the parameter
-        assert hasattr(deepspeed_engine.weight, "ds_id")
+        assert hasattr(model.weight, "ds_id")
+
+    def test_initialize_inside_init(self):
+        ds_config = dict(train_batch_size=1, zero_optimization=dict(stage=3))
+
+        with deepspeed.zero.Init(config_dict_or_path=ds_config):
+            assert (deepspeed.zero.partition_parameters.zero_init_enabled == 1)
+            model = torch.nn.Linear(4, 4)
+            _, *_ = deepspeed.initialize(model=model, config_params=ds_config)
```

Comment:
I see two issues here: 
1.  `deepspeed.initialize()` should not modify init_context in an externally visible way. For example, `deepspeed.zero.partition_parameters.zero_init_enabled` should be the same before and after. 
2. The following test fails with infinite recursion error.
```python
import torch 
import deepspeed 

ds_config = dict(train_batch_size=1, zero_optimization=dict(stage=3))
with deepspeed.zero.Init(config_dict_or_path=ds_config):
    with deepspeed.zero.Init(config_dict_or_path=ds_config):
        model = torch.nn.Linear(4, 4)
        _, *_ = deepspeed.initialize(model=model, config_params=ds_config)
```

Ideally, `deepspeed.initialize` should temporarily disable any existing init_context and re-enable afterwards.

Reply:
I think I understand. I see at least two ways to deal with this:

1. The way it works now - hard shut down the `Init()` context in `initialize()`. This way the partitioning hook is installed in `DeepSpeedEngine` and then removed in `initialize()`. This seems to me to be brittle - if ever `DeepSpeedEngine` needs to be created outside `initialize()` it always needs to be accompanied by `shutdown_init_context()` for reasons that are not immediately obvious. Perhaps this is fine, I defer to your expertise here.
2. I tried explicitly excepting DeepSpeedEngine from the partitioning hook in partition_parameters.py:

```python
    def _enable_class(cls):
        if cls == deepspeed.runtime.engine.DeepSpeedEngine:
            return
        cls._old_init = cls.__init__
        cls.__init__ = partition_after(cls.__init__)

    def _disable_class(cls):
        if cls == deepspeed.runtime.engine.DeepSpeedEngine:
            return
        cls.__init__ = cls._old_init
```

This looks like it works. Maybe it would be better to make a class decorator that excludes classes from partitioning and decorate `DeepSpeedEngine` with it? Then you never have to call `shutdown_init_context()` explicitly.

Reply:
Hi @eisene, thank you for your great improvements. This PR significantly simplifies the management of `zero.Init()` contexts while it allows us to flexibly declare a new model class inside `zero.Init()`

Since the current code in this PR throws in some cases as pointed by @tjruwase, I submitted a separate PR #3592.
This can run the code shown by @tjruwase without an error, and also restores the context of  `zero.Init()` after `deepspeed.initialize()`.

I would be happy if we could discuss with you all including @tjruwase and @stas00.